### PR TITLE
[RFC] Filter admin access

### PR DESCRIPTION
**Pull request by MadeByMike**
Added a filter option to control access to the adminUI. I still think a hook is needed as per #1568 but this is a separate issue. 

I'm not the best with Express routes and it always does my head in a little so if someone has the time to suggest a neater approach I'd be much appreciated.

Still to do is docs, changeset and everything...

Looking for feedback on general idea.

```
module.exports = {
  keystone,
  apps: [
    new GraphQLApp(),
    new AdminUIApp({
      authStrategy: adminAuthStrategy,
      filterAdminAccess: user => {
        console.log({ user });
        return false;
      }
    })
  ]
};
```

### CODE REVIEW

**Comment by MadeByMike** [comment]
This has been working really great for me. If I've got in principal approval I'll add docs and a changeset.

**Comment by MadeByMike** [comment]
> This is probably a "let's fix it later" thing but from what I understand from this code, if a user doesn't have admin access, they'll be redirected to a completely empty page which I imagine would be pretty confusing, could we have a "you don't have permission to access the admin ui" page or something?

Totally agree!

File: packages/app-admin-ui/index.js [hunk 1]
```diff
@@ -52,6 +54,15 @@ class AdminUIApp {
     };
   }
 
+  isAccessAllowed(req) {
+    return (
+      req.user &&
+      this._isAccessAllowed({ authentication: { item: req.user, listKey: req.authedListKey } }) &&
```
Comment (jesstelford) [inline packages/app-admin-ui/index.js]: I made this API consistent with the access control APIs

File: packages/app-admin-ui/index.js [hunk 2]
```diff
@@ -52,6 +54,15 @@ class AdminUIApp {
...
+      req.session.audiences &&
+      req.session.audiences.includes('admin')
```
Comment (jesstelford) [inline packages/app-admin-ui/index.js]: This is potentially redundant now? @timleslie this was part of your initial experimental audiences work - is it useful to keep?

File: packages/app-admin-ui/index.js [hunk 3]
```diff
@@ -149,22 +185,26 @@ class AdminUIApp {
     }
     const secureBuiltRoot = path.join(builtAdminRoot, 'secure');
     const secureStaticMiddleware = express.static(secureBuiltRoot);
-    const secureFallbackMiddleware = fallback('index.html', { root: secureBuiltRoot });
+    const secureFallbackMiddleware = fallback('index.html', {
+      root: secureBuiltRoot,
+    });
```
Comment (jesstelford) [inline packages/app-admin-ui/index.js]: Thanks prettier ¯\\\_(ツ)\_/¯

File: packages/app-admin-ui/index.js [hunk 4]
```diff
@@ -129,16 +138,43 @@ class AdminUIApp {
...
+      // All unauthenticated requests go to the signin page
+      res.format({
+        // For everything else, we have middleware to handle it later down the
+        // line
+        // From the docs: "If the header is not specified [or is */*], the first
+        // callback is invoked."
+        default: () => {
+          next();
+        },
+        // For everything else, we have middleware to handle it later down the
```
Comment (MadeByMike) [inline packages/app-admin-ui/index.js]: Did you mean to copy this comment?

File: packages/app-admin-ui/index.js [hunk 5]
```diff
@@ -176,33 +216,36 @@ class AdminUIApp {
...
 
     _app.use(this.adminPath, app);
 
     if (this.enableDefaultRoute) {
       // Attach this last onto the root so the `this.adminPath` can overwrite it
       // if necessary
-      _app.get('/', (req, res) => res.sendFile(path.resolve(__dirname, './server/default.html')));
+      _app.get('/', (req, res) => {
+        if (!this.isAccessAllowed(req)) {
+          res.sendFile(path.resolve(__dirname, './server/default.html'));
```
Comment (jesstelford) [inline packages/app-admin-ui/index.js]: We should still send this even when the user is allowed access; if they deployed with the defaults of `/admin`, but have `/` bookmarked, they'll then be shown a blank screen, for example.

File: packages/app-admin-ui/index.js [hunk 6]
```diff
@@ -244,13 +287,18 @@ class AdminUIApp {
       const publicHotMiddleware = webpackHotMiddleware(publicCompiler, webpackHotMiddlewareConfig);
 
       // app.use(adminMiddleware);
```
Comment (MadeByMike) [inline packages/app-admin-ui/index.js]: Should this old comment go?

File: packages/app-admin-ui/index.js (hunk 2 above)
Comment (timleslie) [reply packages/app-admin-ui/index.js]: Um, I'd prefer to keep it for now and either get it all to work in the future or tear it all out in one big hit if we decide its redundant.

**Comment by jesstelford** [comment]
Looks like this:

admin-filter

File: packages/app-admin-ui/index.js [hunk 7]
```diff
@@ -176,33 +216,36 @@ class AdminUIApp {
...
     }
 
     _app.use(this.adminPath, app);
 
     if (this.enableDefaultRoute) {
       // Attach this last onto the root so the `this.adminPath` can overwrite it
       // if necessary
-      _app.get('/', (req, res) => res.sendFile(path.resolve(__dirname, './server/default.html')));
+      _app.get('/', (req, res) => {
+        if (!this.isAccessAllowed(req)) {
```
Comment (MadeByMike) [inline packages/app-admin-ui/index.js]:
```suggestion
    
```

File: packages/app-admin-ui/index.js [hunk 8]
```diff
@@ -176,33 +216,36 @@ class AdminUIApp {
...
+          res.sendFile(path.resolve(__dirname, './server/default.html'));
+        }
```
Comment (MadeByMike) [inline packages/app-admin-ui/index.js]:
```suggestion
        
```

File: packages/app-admin-ui/index.js (hunk 5 above)
Comment (jesstelford) [inline packages/app-admin-ui/index.js]:
```suggestion
        res.sendFile(path.resolve(__dirname, './server/default.html'));
```

File: packages/app-admin-ui/index.js (hunk 4 above)
Comment (jesstelford) [reply packages/app-admin-ui/index.js]: Whoops, nope. Good catch!

File: packages/app-admin-ui/index.js (hunk 6 above)
Comment (jesstelford) [reply packages/app-admin-ui/index.js]: 💥

**Pull request merged**